sql: fix cornercases with LIMIT and OFFSET

RaduBerinde opened the pull request:
We fix a few bugs around LIMIT and OFFSET, mostly around overflows. We
also remove the optimization in the distsql planner to return an empty
plan in some cases (this increases the testing surface and should
rarely be useful, as the optimizer can do this optimization when the
values are not subqueries).

Release note (bug fix): fixed internal errors related to very large
LIMIT and/or OFFSET values.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56672)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56672)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/colexec/colbuilder/execplan.go, line 1344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56672#-MMHWbNHEEPCLr-GvgsA:-MMHWbNHEEPCLr-GvgsB:b5u7xsr) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f6373e801cd0b7b23672eb1080e5d09813fc13/pkg/sql/colexec/colbuilder/execplan.go#L1344)):*
> ```Go
> 			offset = maxInt
> 		}
> 		r.Op = colexec.NewOffsetOp(r.Op, int(offset))
> ```

I think we should change `OffsetOp` and `LimitOp` to operate on `uint64` instead to be in line with the post-processing spec.
___
*[pkg/sql/logictest/testdata/logic_test/limit, line 184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56672#-MMHWoJ00ps0cYfOgJul:-MMHWoJ00ps0cYfOgJum:bhe7vzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f6373e801cd0b7b23672eb1080e5d09813fc13/pkg/sql/logictest/testdata/logic_test/limit#L184)):*
> ```
> statement ok
> CREATE TABLE vals (k STRING PRIMARY KEY, v INT);
> INSERT INTO vals VALUES ('zero',0), ('one',1), ('large', 9223372036854775806), ('maxint64', 9223372036854775807);
> ```

nit: missing spaces before `0` and `1`.
___
*[pkg/sql/logictest/testdata/logic_test/limit, line 306 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56672#-MMHX2xz9cpIbZm_PsTt:-MMHX2xz9cpIbZm_PsTu:bexjv3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f6373e801cd0b7b23672eb1080e5d09813fc13/pkg/sql/logictest/testdata/logic_test/limit#L306)):*
> ```
> ----
> 
> # Offset maxin64.
> ```

nit: `s/maxin64/maxint64/`.
___
*[pkg/sql/rowexec/sorter.go, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56672#-MMHXiY0AUcCvXcF99TY:-MMHXiY0AUcCvXcF99TZ:b9b8h1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f6373e801cd0b7b23672eb1080e5d09813fc13/pkg/sql/rowexec/sorter.go#L16)):*
> ```Go
> 	"context"
> 	"fmt"
> 	math "math"
> ```

nit: seems redundant.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56672)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/colbuilder/execplan.go, line 1344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56672#-MMHWbNHEEPCLr-GvgsA:-MMHyp_HAvYYKDW42Sm6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f6373e801cd0b7b23672eb1080e5d09813fc13/pkg/sql/colexec/colbuilder/execplan.go#L1344)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think we should change `OffsetOp` and `LimitOp` to operate on `uint64` instead to be in line with the post-processing spec.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 4 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56672)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @yuzefovich)

___
*[pkg/sql/colexec/colbuilder/execplan.go, line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56672#-MMI4iunChOz3eeF9f8j:-MMI4iunChOz3eeF9f8k:bizk8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/80e99d121aa4c2ab96fc5bfaed787b77464e94e5/pkg/sql/colexec/colbuilder/execplan.go#L413)):*
> ```Go
> 			ordering.Columns, int(matchLen),
> 		)
> 	} else if post.Limit != 0 && post.Filter.Empty() && sumFitsInInt(post.Limit, post.Offset) {
> ```

It'd also be great if the code here was similar to `rowexec/sorter` - looks like we'll only need to change `k` to be of `uint64` type for `NewTopKSorter`, right?


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: fix cornercases with LIMIT and OFFSET

We fix a few bugs around LIMIT and OFFSET, mostly around overflows. We
also remove the optimization in the distsql planner to return an empty
plan in some cases (this increases the testing surface and should
rarely be useful, as the optimizer can do this optimization when the
values are not subqueries).

Release note (bug fix): fixed internal errors related to very large
LIMIT and/or OFFSET values.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56672)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/colbuilder/execplan.go, line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56672#-MMI4iunChOz3eeF9f8j:-MMIsyWB0cygrAfit1Sl:b-8y6r4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/80e99d121aa4c2ab96fc5bfaed787b77464e94e5/pkg/sql/colexec/colbuilder/execplan.go#L413)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

It'd also be great if the code here was similar to `rowexec/sorter` - looks like we'll only need to change `k` to be of `uint64` type for `NewTopKSorter`, right?
</blockquote></details>

We'd still need to check for overflow of the sum. I don't know, here I think we should probably have a value above which we don't use sortTopK - it is likely much slower when we don't get significantly more rows than a limit. I added a TODO to that effect.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56672#-:-MMIzzImAw8bZn5l8baA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/colexec/colbuilder/execplan.go, line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56672#-MMI4iunChOz3eeF9f8j:-MMIzg9sAqlulD86BS66:b-tj6abb) ([raw file](https://github.com/cockroachdb/cockroach/blob/80e99d121aa4c2ab96fc5bfaed787b77464e94e5/pkg/sql/colexec/colbuilder/execplan.go#L413)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We'd still need to check for overflow of the sum. I don't know, here I think we should probably have a value above which we don't use sortTopK - it is likely much slower when we don't get significantly more rows than a limit. I added a TODO to that effect.
</blockquote></details>

Ok. I was thinking that `post.Limit <= math.MaxUint64-post.Offset ` would be sufficient to perform the check if we switched top K to work with uint64 too.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56672)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/colexec/colbuilder/execplan.go, line 413 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56672#-MMI4iunChOz3eeF9f8j:-MMJRvgd6YXu4asYOjHY:b-fp1swd) ([raw file](https://github.com/cockroachdb/cockroach/blob/80e99d121aa4c2ab96fc5bfaed787b77464e94e5/pkg/sql/colexec/colbuilder/execplan.go#L413)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Ok. I was thinking that `post.Limit <= math.MaxUint64-post.Offset ` would be sufficient to perform the check if we switched top K to work with uint64 too.
</blockquote></details>

You're right, it is cleaner, done.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
